Manage users in Everest (#127)
percona/everest-doc: docs/manage_users.md

# Manage users in Percona Everest

Percona Everest provides user management capabilities, enabling you to securely log in through either the Percona Everest UI or the API.

Local user management involves administering Percona Everest users to ensure secure access to database resources. This encompasses tasks such as creating and deleting users, updating their passwords, etc.

When you install Percona Everest, an `admin` user is automatically created, granting full access to the system.

!!! note alert alert-primary "Note"

>     It is recommended that you use the admin account for the initial configuration and subsequently transition to Single Sign-On (SSO) integration.

Reviewer: Would be nice to link a reference to the SSO documentation page, once added

Author: Sure.

Author: Done

      users.yaml: YWRtaW46CiAgcGFzc3dvcmRIYXNoOiBhZG1pbgogIGVuYWJsZWQ6IHRydWUKICBjYXBhYmlsaXRpZXM6CiAgICAtIGxvZ2lu
    ```

Decoding the base64 encoded value will yield a YAML that appears as follows:

??? example "YAML after decoding"

    ```{.text .no-copy}
    admin:
        passwordHash: <password>
    enabled: true
    capabilities:
        - login
    ```

## User management commands

This section provides a list of CLI commands for managing users in Percona Everest.

### Retrieve password

You can retrieve the password for the `admin` user by running the following command:

```sh
everestctl accounts initial-admin-password
```

!!! note alert alert-primary "Note"

    The passwords in this Secret are stored as hashes. However, the initial admin user has the password stored as plain text for convenient retrieval later on. We strongly recommend that you update the admin password after installation.

### Update the password

To update the password for an existing user:

```sh
everestctl accounts set-password -u <username> --new-password $PASSWORD
```

### Create a new user

To create a new user:

```sh
everestctl accounts create -u <username> -p $PASSWORD
```

> This user will have all the permissions to the system.

Reviewer: Please note that this is true only until RBAC is added. After that, new users will have no permissions, and the Everest admin will have to explicitly add new permissions for new users.

Author: @mayankshah1607 In that case should we add a note?

Reviewer: Yes, that would be nice. We can replace this line with a note that says:

```
As of v1.0.0, newly added users have full access to the system. Once RBAC support is added, Everest admins will be able to manage permissions for users, granting them fine-grained control over database resources.
```

(And once RBAC is in, we should also remember to remove/update this note)

Author: ok. Sure.

Author: Fixed.

# Manage users in Percona Everest

Percona Everest provides user management capabilities, enabling you to securely log in through either the Percona Everest UI or the API.

Local user management involves administering Percona Everest users to ensure secure access to database resources. This encompasses tasks such as creating and deleting users, updating their passwords, etc.

When you install Percona Everest, an `admin` user is automatically created, granting full access to the system.

!!! note alert alert-primary "Note"

>     It is recommended that you use the admin account for the initial configuration and subsequently transition to Single Sign-On (SSO) integration.

Reviewer 2: I'm wondering if we shouldn't be more explicit in mentioning that it's not just the `admin account` but the entire local user management in general :shrug:. For production use we recommend SSO.

Author: Done

        
    ```{.text .no-copy}
    apiVersion: v1
    kind: Secret
    metadata:
        name: everest-accounts
        namespace: everest-system
    data:                    
      users.yaml: YWRtaW46CiAgcGFzc3dvcmRIYXNoOiBhZG1pbgogIGVuYWJsZWQ6IHRydWUKICBjYXBhYmlsaXRpZXM6CiAgICAtIGxvZ2lu
    ```

Decoding the base64 encoded value will yield a YAML that appears as follows:

??? example "YAML after decoding"

    ```{.text .no-copy}
    admin:
        passwordHash: <password>
    enabled: true
    capabilities:
        - login
    ```

## User management commands

This section provides a list of CLI commands for managing users in Percona Everest.

### Retrieve password

You can retrieve the password for the `admin` user by running the following command:

```sh
everestctl accounts initial-admin-password
```

!!! note alert alert-primary "Note"

    The passwords in this Secret are stored as hashes. However, the initial admin user has the password stored as plain text for convenient retrieval later on. We strongly recommend that you update the admin password after installation.

### Update the password

To update the password for an existing user:

```sh

> everestctl accounts set-password -u <username> --new-password $PASSWORD

Reviewer 2: @mayankshah1607 if we don't specify the `--new-password $PASSWORD` flag, users will get prompted through the wizard, correct? 
If so maybe should we advertise the version without the `--new-password $PASSWORD` flag so that users get prompted through the wizard and avoid leaking passwords in their shell's history. WDYT?

Reviewer: Yeah, that sounds good.. Let's remove the `--new-password` flag

Author: Done

```

### Create a new user

To create a new user:

```sh

> everestctl accounts create -u <username> -p $PASSWORD

Reviewer 2: Same as previous comment

Author: Done

```

!!! info "Important"
    For Percona Everest version 1.0.0, new users have full access to the system. However, once RBAC support is in place, an admin user will be able to manage permissions for users, granting them fine-grained control over database resources.

### List the users

To list all the users in Percona Everest:

```sh
everestctl accounts list
```

### Delete a user

To delete an existing user:

```sh
everestctl accounts delete -u <username>
```

## JSON Web Token (JWT) and keys

When you log in from the UI, Percona Everest issues a JSON Web Token to authenticate the requests. By default, this token is valid for **24** hours, after which you are expected to log in again.

!!! info "Important"

    Since JWT authentication is stateless, it is currently impossible to explicitly revoke specific tokens. Therefore, even a deleted user may continue to request the API as long as they have a valid token.

> Everest uses the RSA algorithm to sign in and verify JWT. The RSA keys used in  the backend are automatically generated upon installation and stored in the `everest-jwt` Secret in the `everest-system` namespace.

Reviewer 2: ```suggestion
The Everest API uses the RSA algorithm to sign and verify the JWT. The RSA key pair used for this is automatically generated upon installation and stored in the `everest-jwt` Secret in the `everest-system` namespace.
```

        apiVersion: v1
        data:
            id_rsa: <PRIVATE KEY>
            id_rsa.pub: <PUBLIC KEY>
        kind: Secret
        metadata:
            name: everest-jwt
            namespace: everest-system
        type: Opaque

> To reset the keys:

Reviewer 2: ```suggestion
To reset the key pair:
```
